package/MDAnalysis/analysis/diffusionmap.py

```diff
@@ -0,0 +1,383 @@
+    -------
+    save(filename)
+        Save the `dist_matrix` to a given filename
+    """
+    def __init__(self, u, select='all', metric=rms.rmsd, cutoff=1E0-5,
+                 weights=None, start=None, stop=None, step=None):
+        """
+        Parameters
+        ----------
+        u : trajectory `~MDAnalysis.core.AtomGroup.Universe`
+            The MD Trajectory for dimension reduction, remember that
+            computational cost of eigenvalue decomposition
+            scales at O(N^3) where N is the number of frames.
+            Cost can be reduced by increasing step interval or specifying a
+            start and stop.
+        select: str, optional
+            Any valid selection string for
+            :meth:`~MDAnalysis.core.AtomGroup.AtomGroup.select_atoms`
+            This selection of atoms is used to calculate the RMSD between
+            different frames. Water should be excluded.
+        metric : function, optional
+            Maps two numpy arrays to a float, is positive definite and
+            symmetric, Default: metric is set to rms.rmsd().
+        cutoff : float, optional
+            Specify a given cutoff for metric values to be considered equal,
+            Default: 1EO-5
+        weights : array, optional
+            Weights to be given to coordinates for metric calculation
+        start : int, optional
+            First frame of trajectory to analyse, Default: 0
+        stop : int, optional
+            Last frame of trajectory to analyse, Default: -1
+        step : int, optional
+            Step between frames to analyse, Default: 1
+        """
+
+        self._u = u
+        traj = self._u.trajectory
+        self.atoms = self._u.select_atoms(select)
+        self._metric = metric
+        self._cutoff = cutoff
+        self._weights = weights
+        # remember that this must be called before referencing self.nframes
+        self._setup_frames(traj, start, stop, step)
+
+    def _prepare(self):
+        logger.info('numframes {0}'.format(self.nframes))
+        self.dist_matrix = np.zeros((self.nframes, self.nframes))
+
+    def _single_frame(self):
+        traj_index = self._ts.frame
+        i_ref = self.atoms.positions - self.atoms.center_of_mass()
+        # diagonal entries need not be calculated due to metric(x,x) == 0 in
+        # theory, _ts not updated properly. Possible savings by setting a
+        # cutoff for significant decimal places to sparsify matrix
+        for j in range(self.stop-1, self._ts.frame-1, -self.step):
+            self._ts = self._u.trajectory[j]
+            j_ref = self.atoms.positions-self.atoms.center_of_mass()
+            dist = self._metric(i_ref, j_ref, weights=self._weights)
+            # distance squared in preparation for kernel calculation
+            # don't think this will come up in other areas, but might be
+            # a fix we should make later
+            self.dist_matrix[traj_index, j] = dist**2 if dist > self._cutoff else 0
```

Comment:
Oh! Good catch! I think I was in a bit of a zone and reading the Lafon
paper and copying that when we're using RMSD not Euclidean distance. I'll
fix that.
On Fri, Jun 17, 2016 at 12:48 AM kain88-de <notifications@github.com> wrote:

> In package/MDAnalysis/analysis/diffusionmap.py
> <https://github.com/MDAnalysis/mdanalysis/pull/863#discussion_r67471628>:
>
> > +        self.dist_matrix = np.zeros((self.nframes, self.nframes))
> > +
> > +    def _single_frame(self):
> > +        traj_index = self._ts.frame
> > +        i_ref = self.atoms.positions - self.atoms.center_of_mass()
> > +        # diagonal entries need not be calculated due to metric(x,x) == 0 in
> > +        # theory, _ts not updated properly. Possible savings by setting a
> > +        # cutoff for significant decimal places to sparsify matrix
> > +        for j in range(self.stop-1, self._ts.frame-1, -self.step):
> > +            self._ts = self._u.trajectory[j]
> > +            j_ref = self.atoms.positions-self.atoms.center_of_mass()
> > +            dist = self._metric(i_ref, j_ref, weights=self._weights)
> > +            # distance squared in preparation for kernel calculation
> > +            # don't think this will come up in other areas, but might be
> > +            # a fix we should make later
> > +            self.dist_matrix[traj_index, j] = dist**2 if dist > self._cutoff else 0
>
> why are you squaring the distance here? This is changing the used metric.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/MDAnalysis/mdanalysis/pull/863/files/09702a58193e6c0a13dc380235fffb668d5d34ab#r67471628>,
> or mute the thread
> <https://github.com/notifications/unsubscribe/AKcARjPvDNQlovdgyOdGb0VzoRqu8tgNks5qMlEzgaJpZM4IpYIS>
> .
>
-- 
Have a wonderful day,
John J. Detlefs